internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,729 @@
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		workerAuthRepoStorage.workerAuthTransaction = workerTransaction
+	}
+
+	return workerAuthRepoStorage, nil
+}
+
+func createTransaction(ctx context.Context, w db.Writer) (*WorkerAuthTransaction, error) {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).createTransaction"
+	txWriter, transaction, err := w.BeginTx(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	workerTransaction := &WorkerAuthTransaction{
+		transactionWriter: txWriter,
+		transaction:       transaction,
+	}
+
+	return workerTransaction, nil
+}
+
+// Flush is called when storage is done being performed. The boolean parameter
+// indicates whether the operation was successful (true) or failed (false).
+func (r *WorkerAuthRepositoryStorage) Flush(b bool) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).Flush"
+	if r.workerAuthTransaction == nil {
+		return nil
+	}
+	if b == false {
+		err := r.workerAuthTransaction.transaction.Rollback(nil)
+		if err != nil {
+			return fmt.Errorf("(%s) Transaction rollback failed: %w", op, err)
+		}
+	} else {
+		err := r.workerAuthTransaction.transaction.Commit(nil)
+		if err != nil {
+			return fmt.Errorf("(%s) Transaction commit failed: %w", op, err)
+		}
+	}
+	return nil
+}
+
+// Store implements the Storage interface
+func (r *WorkerAuthRepositoryStorage) Store(ctx context.Context, msg nodee.MessageWithId) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).Store"
+	if err := types.ValidateMessage(msg); err != nil {
+		return fmt.Errorf("(%s) given message cannot be stored: %w", op, err)
+	}
+	if msg.GetId() == "" {
+		return fmt.Errorf("(%s) given message cannot be stored as it has no ID.", op)
+	}
+
+	// Determine type of message to store
+	marshaledBytes, err := proto.Marshal(msg.(proto.Message))
+	if err != nil {
+		return fmt.Errorf("error marshaling nodee.MessageWithId: %w", err)
+	}
+	switch msg.(type) {
+	case *types.NodeInformation:
+		node, err := unmarshalNodeInformation(ctx, marshaledBytes)
+		if err != nil {
+			return err
+		}
+		err = r.storeNodeInformation(ctx, node)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	case *types.RootCertificate:
+		cert, err := unmarshalRootCertificate(ctx, marshaledBytes)
+		if err != nil {
+			return err
+		}
+		err = r.storeRootCertificate(ctx, cert)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	default:
+		err = fmt.Errorf("(%s) Message type not supported for Store", op)
+		return errors.Wrap(ctx, err, op)
+	}
+
+	return nil
+}
+
+// Node information is stored in two parts:
+// * the workerAuth record is stored with a reference to a worker
+// * certificate bundles are stored with a reference to the workerAuth record and issuing root certificate
+func (r *WorkerAuthRepositoryStorage) storeNodeInformation(ctx context.Context, node *types.NodeInformation) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).storeNodeInformation"
+
+	if r.worker == nil {
```

Comment:
The first time node information is being stored I don't know for sure you'd have a worker yet. This is part of the discussions that happened today around when a worker and its ID are created but it may be that as part of storing node information worker creation will have to happen first, so either that'd have to happen and this would then not be nil, or this being nil would trigger that.

Reply:
I'd been assuming the calling function would handle worker creation; this nil check was in case the caller didn't pass in a worker when attempting to perform nodeInformation storage. 

I think worker lookup/ creation should happen first and should be external, esp when we get to worker rekeying.